abgc/src/lib.rs

```diff
@@ -30,75 +29,101 @@ pub trait GcLayout {
 
 #[derive(Debug)]
 pub struct Gc<T: GcLayout> {
-    ptr: *mut GcBox<T>,
+    objptr: *mut T,
 }
 
 impl<T: GcLayout> Gc<T> {
     /// Create a `Gc` from `v`.
     pub fn new(v: T) -> Self {
-        let (gcbptr, vptr): (_, *mut T) = GcBox::<T>::alloc_blank(Layout::new::<T>());
+        let objptr = Gc::<T>::alloc_blank(Layout::new::<T>());
         let gc = unsafe {
-            vptr.copy_from_nonoverlapping(&v, 1);
-            Gc::from_raw(gcbptr)
+            objptr.copy_from_nonoverlapping(&v, 1);
+            Gc::from_raw(objptr)
         };
         forget(v);
         gc
     }
 
+    /// Allocate memory sufficient to `l` (i.e. correctly aligned and of at least the required
+    /// size). The returned pointer must be passed to `Gc::from_raw`.
+    pub fn alloc_blank(l: Layout) -> *mut T {
+        let (layout, uoff) = Layout::new::<usize>().extend(l).unwrap();
+        // In order for our storage scheme to work, it's necessary that `uoff - sizeof::<usize>()`
+        // gives a valid alignment for a `usize`. There are only two cases we need to consider
+        // here:
+        //   1) `object`'s alignment is smaller than or equal to `usize`. If so, no padding will be
+        //      added, at which point by definition `uoff - sizeof::<usize>()` will be exactly
+        //      equivalent to the start point of the layout.
+        //   2) `object`'s alignment is bigger than `usize`. Since alignment must be a power of
+        //      two, that means that we must by definition be adding at least one exact multiple of
+        //      `usize` bytes of padding.
```

Comment:
If I understand correctly, this large comment boils down to: `Layout::extend()` guarantees the correct alignment.

Am I right?

Reply:
Yes (https://doc.rust-lang.org/std/alloc/struct.Layout.html#method.extend has the gory details).

Reply:
Right. So counter-intuitively, the comment led me to believe something much more clever was happening than actually was.

Why don't we keep the assertion, but remove the large comment? The invariants of `extend()` are documented elsewhere and I don't think it's necessary to duplicate.